umpf

```diff
@@ -1855,6 +1864,197 @@ do_distribute() {
+			sed 's/\s\+.*$//')
+	else
+		commit="refs/tags/$tag"
+	fi
+
+	resolve_commitish "${commit}^{}"
+}
+
+update_local() {
+	local success=false args="${1}"
+	local opts
+
+	${FORCE} && opts+="--force"
+
+	local line
+	while read -r line; do
+		local prefix="" suffix=""
+
+		eval set -- ${line}
+		[ ${#} -eq 0 ] && continue
+
+		local refparsed=$(shorten_commitish $3)
+
+		if [ -z "${refparsed}" ]; then
+			prefix=" * [new branch]"
+		elif [[ "${1}" = *~* ]]; then
+		       prefix="   ${1}..$(shorten_commitish ${2})"
+		elif $FORCE; then
+		       prefix=" + ${refparsed}..$(shorten_commitish ${2})"
+		       suffix="  (forced update)"
+		else
+		       prefix=" ! [rejected]"
+		       suffix="  (non-fast-forward)"
+		fi
+
+		printf "%-40s %s -> %s%s\n" "$prefix" $2 $3 "$suffix"
+	done <<< "$args"
+
+	while read -r line; do
+		eval set -- ${line}
+		[ ${#} -eq 0 ] && continue
+
+		if $DRYRUN || git branch $opts $3 $2; then
+			success=true
+		fi
+	done <<< "$args"
+
+	$success || abort
+}
+
+do_push () {
+	local opts args remote
+	local -a branches branch_names
+	local -A topics
+
+	if [ -z "${GIT_REMOTE}" ]; then
+		info "Git remote must be specified. Cannot continue."
+		exit 1
+	fi
+
+	if [ "${GIT_REMOTE}" != "refs/heads/" ]; then
+		remote=${GIT_REMOTE%/}
+	fi
+
+	prepare_persistent push "${@}"
+	parse_series push "${STATE}/series"
+
+	local tagname="$(<"${STATE}/tagname")"
+	local tagrevf="$(<"${STATE}/tagrev-flat")"
+	mapfile -t branches < "${STATE}/topics"
+	mapfile -t branch_names < "${STATE}/topic-names"
+
+	if [ -n "${remote}" ]; then
+		# Needed, so git rev-parse below can check for existent branches
+		git fetch --quiet --no-tags ${remote} 2>/dev/null
+	fi
+
+	local rtagrev="$(resolve_tag "${remote}" ${tagname})"
+	local rtagrevf="$(resolve_commitish ${rtagrev}^)"
+	if [ "$tagrevf" != "$rtagrevf" ]; then
+		if [ -z "$rtagrevf" ]; then
+			abort "${remote}${remote:+/}refs/tags/$tagname not found"
+		else
+			abort "${remote}${remote:+/}refs/tags/$tagname" \
+			      "has unexpected commit-ish $rtagrev"
+		fi
+	fi
+
+	for i in "${!branch_names[@]}"; do
+		local branch=${branch_names[$i]}
+		local rbranchrev="$(resolve_commitish "${GIT_REMOTE}${branch}")"
+
+		[ -z "${remote}" ] && [ "${branches[$i]}" = "${rbranchrev}" ] && continue
+
+		# Don't touch local branches that are already on the correct revision.
+		# For remote branches, we let git push handle it.
+		[ -z "${remote}" ] && [ "${branches[$i]}" = "${rbranchrev}" ] && continue
+		$UPDATE && [ -z "${rbranchrev}" ] && continue
+
+		topics[${branch}]=$(resolve_commitish ${branches[$i]})
+	done
+
+	if [ -n "${remote}" ]; then
+		${FORCE}  && opts+="--force-with-lease"
+		${DRYRUN} && opts+="--dry-run"
+
+		for topic in "${!topics[@]}"; do
+			args+="${topics[$topic]}:refs/heads/${topic} "
+		done
+
+		if [ -z "$args" ]; then
+			info "No branches to push"
+			cleanup
+			return
+		fi
+
+		${GIT} push $opts ${remote} -- $args
```

Comment:
I think, if you use `${GIT_DIR}` as remote for the local case, you can use `git push` for that as well. That would avoid a lot of complexity. So try that first, it may make some of my comments obsolete.

Reply:
Nice, switched over to that.